### PR TITLE
[os] fix #10017 regression, fix #10025 regression

### LINKED ISSUE #10017: regression: normalizePath("foo/..") now incorrectly returns `""`, should be `"."` like before + in almost all other languages
/cc @Araq 
my code broke due to a recent regression introduced here: ce9815bcf5a0ae5549be116604c2ad8ddacb17af
 
```nim
import os
block:
  var s = "foo/.."
  normalizePath(s)
  echo "{", s, "}" # WAS {.}; now: {}
```

## other languages:
conflating `.` with empty string is a bad idea; empty string (as a path) represents un-initialized path; this is different from `.` which represents `current dir`

Pretty much all other languages I use don't do that mistake:
* D: buildNormalizedPath("foo/..") prints `.`
* python: os.path.normpath('foo/..') prints `.`
* C++: boost::filesystem::path("foo/..").lexically_normal().string() => `.`
* node JS: path.normalize('foo/..') prints `.`
* go:  fmt.Println(filepath.Clean("foo/..")) prints `.`

## note
there was already a test case in place:
doAssert normalizedPath(".") == "."
but it got overwritten by ce9815bcf5a0ae5549be116604c2ad8ddacb17af
`doAssert normalizedPath(".") == ""`

[EDIT]
see also @krux02 's comment from https://github.com/nim-lang/Nim/issues/10017#issuecomment-447877313:
> you forgot the most important example, the shell. No matter if it is Linux or Window, everywhere the current directory is ".". "" is not a valid path on my shell (fish).

### LINKED ISSUE #10025: [regression] [os]  joinPath("usr/", "/lib")="usr/lib" but  joinPath(".", "/lib")="/lib"
/cc @Araq 

```nim
echo joinPath("usr/", "/lib")
echo joinPath(".", "/lib")
echo joinPath("./", "/lib")
```
now prints

usr/lib
/lib
/lib

which is definitely wrong (inconsistent, regardless of which side of https://github.com/nim-lang/Nim/issues/8268 you're on)

the bug was introduced here: 9cc4a57768156883acc4e8190c78a69ed2c3d526 /cc @Araq 

before which it was printing:
usr/lib
./lib
./lib

## note
I had argued here https://github.com/nim-lang/Nim/issues/8268 that `joinPath("foo", "/bar")` should've returned "/bar" as it's the sane thing to do (and as done in most languges), but not sure whether we want to introduce this breaking change now; In this PR (sent out a while ago...) https://github.com/nim-lang/Nim/pull/8351 I've introduced instead an extra param to avoid this breaking change:
```
doAssert joinPath("foo", "/bar", "/baz", "tail", absOverrides = true) == "/baz/tail"
```
maybe now's a good time to revive it

* adding high priority because regressions like this can cause a lot of harm (eg wrong file being deleted etc)

### PR DESCRIPTION
**Pull request by timotheecour**
/cc @Araq @krux02 @dom96
* fix #10017 regression
* fix #10025 regression

### CODE REVIEW

**Comment by timotheecour** [comment]
> But then what does "" mean? And why does "" not compare equal to "."?

`""` means un-initialized. It should never ever mean "current directory"

try:
```
ls ""
ls: cannot access '': No such file or directory
```

it will be a source of endless bugs.
see also @krux02 's comment https://github.com/nim-lang/Nim/issues/10017#issuecomment-447877313

File: lib/pure/pathnorm.nim [hunk 1]
```diff
@@ -100,3 +100,4 @@ proc normalizePath*(path: string; dirSep = DirSep): string =
   result = newStringOfCap(path.len)
   var state = 0
   addNormalizePath(path, result, state, dirSep)
+  if result == "" and path != "": result = "."
```
Comment (krux02) [inline lib/pure/pathnorm.nim]: `addNormalizePath` is exported, too. But you exclude your fix from that function, why?

Comment (timotheecour) [reply lib/pure/pathnorm.nim]: fixed

**Comment by timotheecour** [comment]
2 birds, one stone! latest commit also fixes #10025

**Comment by Araq** [comment]
Ok, sometimes it's easy to convince me. Thanks a lot!

**Pull request merged**